heroic-core/src/main/java/com/spotify/heroic/cluster/CoreClusterManager.java

```diff
@@ -274,7 +276,33 @@ public void register(LifeCycleRegistry registry) {
             throw new IllegalStateException("Registry not ready");
         }
 
-        return registry.findFromAllShards(OptionalLimit.empty());
+        final List<Pair<Map<String, String>, List<ClusterNode>>> shards =
+            registry.findFromAllShards(OptionalLimit.empty());
+
+        if (expectedShards.isPresent()) {
+            final Set<Map<String, String>> foundShards = new HashSet<>(shards
+                .stream()
+                .map(mapListPair -> mapListPair.getKey())
+                .collect(Collectors.toList()));
+
+            final List<Map<String, String>> shardsWithNoNodes = new ArrayList<>();
+            for (final Map<String, String> expected : expectedShards.get()) {
+                if (!foundShards.contains(expected)) {
+                    shardsWithNoNodes.add(expected);
+                }
+            }
+
+            if (shardsWithNoNodes.size() > 0) {
+                for (final Map<String, String> shard : shardsWithNoNodes) {
+                    /* For every shard that didn't have a discovered node, add an empty entry in the
+                     * shard list. This ensures that suitable code later on will complain that there
+                     * were shards with no available nodes/groups. */
+                    shards.add(Pair.of(shard, ImmutableList.of()));
```

Comment:
Don't modify things returned from other APIs, they could easily be immutable. Instead, copy.